runtime/memory.c

```diff
@@ -148,24 +147,34 @@ Caml_inline void write_barrier(
    }
 }
 
+#if defined(WITH_THREAD_SANITIZER) && defined(NATIVE_CODE)
+#define TSAN_FUNC_ENTRY __tsan_func_entry(__builtin_return_address(0))
+#define TSAN_FUNC_EXIT __tsan_func_exit(NULL)
+#define TSAN_SIGNAL_PLAIN_WRITE(loc) __tsan_write8((void *)loc)
+#else
+#define TSAN_FUNC_ENTRY
+#define TSAN_FUNC_EXIT
+#define TSAN_SIGNAL_PLAIN_WRITE(loc)
+#endif
+
 CAMLno_user_tsan /* We remove the ThreadSanitizer instrumentation of memory
                     accesses by the compiler and instrument manually, because
                     we want ThreadSanitizer to see a plain store here (this is
                     necessary to detect data races). */
 CAMLexport CAMLweakdef void caml_modify (volatile value *fp, value val)
 {
+  TSAN_FUNC_ENTRY;
   write_barrier((value)fp, 0, *fp, val);
 
   /* See Note [MM] above */
   atomic_thread_fence(memory_order_acquire);
-#if defined(WITH_THREAD_SANITIZER) && defined(NATIVE_CODE)
   /* The release store below is not instrumented because of the
    * CAMLno_user_tsan. We signal it to ThreadSanitizer as a plain store (see
    * ocaml-multicore/ocaml-tsan/pull/22#issuecomment-1377439074 on Github).
    */
-  __tsan_write8((void *)fp);
-#endif
```

Comment:
I wonder if the same thing could be written, without the TSAN_* macros above:

```c
#if defined(WITH_THREAD_SANITIZER) && defined(NATIVE_CODE)
  /* The release store below is not instrumented because of the
   * CAMLno_user_tsan. We signal it to ThreadSanitizer as a plain store (see
   * ocaml-multicore/ocaml-tsan/pull/22#issuecomment-1377439074 on Github).
   */
  __tsan_func_entry(__builtin_return_address(0));
  __tsan_write8((void *)fp);
  __tsan_func_exit(NULL);
#endif
```

Reply:
Note: I wonder, why is there no TSAN instrumentation in bytecode here? (I tried to follow the URL pointer but didn't find much information there about this question.)

Reply:
@fabbing pushed a version along the lines of your proposal. However, the `#if` has to be split in two, otherwise the call to `write_barrier` would be made before the `__tsan_func_entry`, which could lead to confusing backtraces in case of a data race there, i.e., the backtrace would miss the function that called `caml_modify`.

The reasons for not instrumenting the bytecode version are given in https://github.com/ocaml-multicore/ocaml-tsan/pull/17. In short, it makes the compiler build very slow (as well as the installation of packages such as dune). Solutions attempted to work around this were sketchy and complexified the build system for no obvious benefit: it is unclear what would be the substantial gain in running bytecode programs with TSan.

Reply:
An additional reason is that the “dynamic” instrumentation of function entries and exits upon raising and effect operations has been implemented for the native backend only, and would require significant work to be implemented in the bytecode interpreter as well. We didn’t want to do that without a strong need for running bytecode programs under TSan.